### PR TITLE
DM-45744 Implement ATBuilding CSC +790-148

### PR DESCRIPTION
**Pull request by bbrondel**

### CODE REVIEW

File: .pre-commit-config.yaml
Comment (tribeiro) [inline .pre-commit-config.yaml:1]: Please, don't commit the `.pre-commit-config.yaml` file. This should be generated by `generate_pre_commit_conf` from `ts-pre-commit-config` package.

File: python/lsst/ts/vent/controller/__init__.py [hunk 1]
```diff
@@ -20,10 +20,12 @@
...
+    from .version import __version__  # Generated by sconsUtils
 except ImportError:
     __version__ = "?"
 
 from .config import Config
-from .controller import Controller, VentGateState
+from .controller import Controller
 from .dispatcher import Dispatcher
+
+__all__ = ["__version__", "Config", "Controller", "Dispatcher"]
```
Comment (tribeiro) [inline python/lsst/ts/vent/controller/__init__.py]: We usually don't have this `__all__`` definition here in the `__init__` file.

File: python/lsst/ts/vent/controller/controller.py [hunk 2]
```diff
@@ -22,33 +22,49 @@
 import logging
 from enum import IntEnum
 
-import megaind
+try:
+    import megaind
```
Comment (tribeiro) [inline python/lsst/ts/vent/controller/controller.py]: I missed the inclusion of this package dependency. I don't think this is part of our stack. Do you really need it? 

Usually, adding new package dependencies require approval by myself and/or Wouter. We should discuss this.

File: python/lsst/ts/vent/controller/controller.py [hunk 3]
```diff
@@ -22,33 +22,49 @@
...
 
 from . import vf_drive
 from .config import Config
+from .simulate import DomeVentsSimulator
 
-__all__ = ["Controller"]
+__all__ = ["Controller", "VentGateState"]
 
 
 class VentGateState(IntEnum):
```
Comment (tribeiro) [inline python/lsst/ts/vent/controller/controller.py]: Should this be in ts-xml?

File: python/lsst/ts/vent/controller/controller.py [hunk 4]
```diff
@@ -22,33 +22,49 @@
...
-    CLOSED = 0
-    PARTIALLY_OPEN = 1
-    OPEN = 2
+    CLOSED = 1
+    PARTIALLY_OPEN = 2
+    OPEN = 3
     FAULT = -1
 
 
+class FanDriveState(IntEnum):
```
Comment (tribeiro) [inline python/lsst/ts/vent/controller/controller.py]: Shouldn't this be in ts-xml?

File: python/lsst/ts/vent/controller/controller.py [hunk 5]
```diff
@@ -22,33 +22,49 @@
...
+
 class Controller:
     """A controller that commands the components associated with the AT dome
     vents and fans. The code in this class is meant to run on the Raspberry Pi
     described in
     https://confluence.lsstcorp.org/display/~fritzm/Auxtel+Vent+Gate+Automation.
     """
 
-    def __init__(self, config: Config = Config()):
+    def __init__(self, config: Config = Config(), simulate: bool = False):
```
Comment (tribeiro) [inline python/lsst/ts/vent/controller/controller.py]: I don't think it is a good idea to pass `config = Config()` as default parameter. In cases like this I would suggest

```
def __init__(self, config: Config | None = None, simulate: bool = False):
    self.config = config if config is not None else Config()
```

File: python/lsst/ts/vent/controller/controller.py [hunk 6]
```diff
@@ -22,33 +22,49 @@
...
         self.cfg = config
```
Comment (tribeiro) [inline python/lsst/ts/vent/controller/controller.py]: Please expand the attribute name to `self.config` instead.

File: python/lsst/ts/vent/controller/controller.py [hunk 7]
```diff
@@ -22,33 +22,49 @@
...
         self.default_fan_frequency = self.cfg.max_freq
         self.log = logging.getLogger(type(self).__name__)
+        self.simulate = simulate
```
Comment (tribeiro) [inline python/lsst/ts/vent/controller/controller.py]: Do you really need to store the `simulate` flag? I think you can determine if it is in simulation mode by checking it `self.simulator is None`.

File: python/lsst/ts/vent/controller/controller.py [hunk 8]
```diff
@@ -209,11 +245,35 @@ async def vfd_fault_reset(self) -> None:
...
+        assert self.connected
         for address, value in vf_drive.FAULT_RESET_SEQUENCE:
             await self.vfd_client.write_register(
                 slave=self.cfg.slave, address=address, value=value
             )
 
+    async def get_drive_state(self) -> FanDriveState:
+        """Returns the current fan drive state based on the contents
+        of the IPAE register.
+
```
Comment (tribeiro) [inline python/lsst/ts/vent/controller/controller.py:267]: Please, document return value.

File: python/lsst/ts/vent/controller/controller.py [hunk 9]
```diff
@@ -209,11 +245,35 @@ async def vfd_fault_reset(self) -> None:
...
+        Raises
+        ------
+        ModbusException
```
Comment (tribeiro) [inline python/lsst/ts/vent/controller/controller.py:273]: It also raises an `AssertionError` it if is not connected.

File: python/lsst/ts/vent/controller/controller.py [hunk 10]
```diff
@@ -209,11 +245,35 @@ async def vfd_fault_reset(self) -> None:
...
+            If a communications error occurs.
+        """
+
+        assert self.connected
+        ipae = (
```
Comment (tribeiro) [inline python/lsst/ts/vent/controller/controller.py:285]: what doesn't `ipae` means? can you expand the acronym?

File: python/lsst/ts/vent/controller/controller.py [hunk 11]
```diff
@@ -231,8 +291,9 @@ async def last8faults(self) -> list[tuple[int, str]]:
         """
 
         self.log.debug("last8faults")
+        assert self.connected
         rvals = await self.vfd_client.read_holding_registers(
-            slave=1, address=vf_drive.Registers.FAULT_REGISTER, count=8
+            slave=self.cfg.slave, address=vf_drive.Registers.FAULT_REGISTER, count=8
```
Comment (tribeiro) [inline python/lsst/ts/vent/controller/controller.py]: I would avoid the use of terms like `slave`. Please, consider renaming this; e.g. client or any other less offensive term.

File: python/lsst/ts/vent/controller/controller.py [hunk 12]
```diff
@@ -246,6 +307,10 @@ def vent_open(self, vent_number: int) -> None:
 
         Raises
         ------
+        ModuleNotFoundError
```
Comment (tribeiro) [inline python/lsst/ts/vent/controller/controller.py]: So, you are not really raising this exception here. But we should talk abut this.

File: python/lsst/ts/vent/controller/controller.py [hunk 13]
```diff
@@ -327,3 +403,41 @@ def vent_state(self, vent_number: int) -> VentGateState:
                 return VentGateState.CLOSED
             case _:
                 return VentGateState.FAULT
+
+    def getOptoCh(self, *args, **kwargs) -> int:
```
Comment (tribeiro) [inline python/lsst/ts/vent/controller/controller.py]: Method should be `snake_case`. But we need to talk about the use of this `megaind` dependency.

File: python/lsst/ts/vent/controller/controller.py [hunk 14]
```diff
@@ -327,3 +403,41 @@ def vent_state(self, vent_number: int) -> VentGateState:
...
+
+        assert self.connected
+        if self.simulate:
+            return self.simulator.getOptoCh(*args, **kwargs)
+        else:
+            if not MEGAIND_AVAILABLE:
+                raise ModuleNotFoundError("The megaind module is not available.")
+            return megaind.getOptoCh(*args, **kwargs)
+
+    def setOd(self, *args, **kwargs) -> None:
```
Comment (tribeiro) [inline python/lsst/ts/vent/controller/controller.py]: Methods should be `snake_case` but, same as above, we need to talk about the use of `megaind`.

File: python/lsst/ts/vent/controller/dispatcher.py [hunk 15]
```diff
@@ -27,7 +49,11 @@ def _cast_string_to_type(new_type: type, value: str):
     """
 
     if new_type is bool:  # Boolean is a special case
-        return value.lower() in ("true", "t", "1")
+        if value.lower() in ("true", "t", "1"):
+            return True
+        if value.lower() in ("false", "f", "0"):
```
Comment (tribeiro) [inline python/lsst/ts/vent/controller/dispatcher.py]: should this be `elif` instead?

File: python/lsst/ts/vent/controller/dispatcher.py [hunk 16]
```diff
@@ -27,7 +49,11 @@ def _cast_string_to_type(new_type: type, value: str):
...
+            return False
+        raise ValueError(f"Expected bool value but got {value}")
```
Comment (tribeiro) [inline python/lsst/ts/vent/controller/dispatcher.py]: Maybe document what are the valid values for conversion to bool?

File: python/lsst/ts/vent/controller/dispatcher.py [hunk 17]
```diff
@@ -142,11 +175,25 @@ async def read_and_dispatch(self) -> None:
                 )
             )
 
-    async def close_vent_gate(self, gate: int) -> None:
-        self.controller.vent_close(gate)
+    async def close_vent_gate(
+        self, gate0: int, gate1: int, gate2: int, gate3: int
+    ) -> None:
+        for gate in (gate0, gate1, gate2, gate3):
+            if gate >= 0 and gate <= 3:
```
Comment (tribeiro) [inline python/lsst/ts/vent/controller/dispatcher.py:194]: This gives me the impression the value of `gate` could be a enum instead. Then you could check it agains valid enum values. But I am not sure I fully understand what it going on here.

File: python/lsst/ts/vent/controller/simulate.py
Comment (tribeiro) [inline python/lsst/ts/vent/controller/simulate.py:1]: Consider renaming this `dome_vents_simulator.py`

**Review by tribeiro — COMMENTED** [review]
Here is an initial round of comments. 

General comment:

Please, document all the assertion Errors that can happen with the inclusion `assert self.connected`.

**Review by tribeiro — CHANGES_REQUESTED** [review]
Forgot to mark "request changes"

File: python/lsst/ts/vent/controller/controller.py (hunk 2 above)
Comment (bbrondel) [reply python/lsst/ts/vent/controller/controller.py]: I think it's necessary as a part of the hardware controlling the vents - the Sequent Microsystems Industrial Automation Hat.

File: python/lsst/ts/vent/controller/dispatcher.py (hunk 17 above)
Comment (bbrondel) [reply python/lsst/ts/vent/controller/dispatcher.py:194]: Yeah, were you involved with that discussion on Slack? Basically  ventOpenGate and ventCloseGate take four ints, and we were loathe to change the CSC if we could avoid that. The result of the discussion was that each gate number passed (0 through 3) would be opened/closed, and -1 would be used as a placeholder.

File: python/lsst/ts/vent/controller/controller.py (hunk 10 above)
Comment (bbrondel) [reply python/lsst/ts/vent/controller/controller.py:285]: It's called the IPAE register in the manual for the ATV320. They don't say what IPAE is supposed to mean or stand for. I've added some clarification in the comments.

File: python/lsst/ts/vent/controller/controller.py (hunk 8 above)
Comment (tribeiro) [reply python/lsst/ts/vent/controller/controller.py:267]: Please use the [numpy docs style](https://numpydoc.readthedocs.io/en/latest/format.html#returns).

This should be something like:

```
Returns
--------
`FanDriveState`
    Current fan drive state.
```

You don't need to document all the values as that should be in the enumeration.

File: python/lsst/ts/vent/controller/controller.py [hunk 18]
```diff
@@ -302,28 +403,67 @@ def vent_state(self, vent_number: int) -> VentGateState:
...
                 return VentGateState.PARTIALLY_OPEN
             case 0, 1:
                 return VentGateState.CLOSED
             case _:
                 return VentGateState.FAULT
+
+    def get_opto_ch(self, *args, **kwargs) -> int:
+        """Calls hardware I/O or a simulated substitute depending
+        whether the class was instantiated with simulate = True.
+
```
Comment (tribeiro) [inline python/lsst/ts/vent/controller/controller.py:436]: Please, document input arguments and return values (use numpy docs style).

File: python/lsst/ts/vent/controller/controller.py [hunk 19]
```diff
@@ -302,28 +403,67 @@ def vent_state(self, vent_number: int) -> VentGateState:
...
+        assert self.connected
+        if self.simulator is not None:
+            return self.simulator.get_opto_ch(*args, **kwargs)
+        else:
+            raise NotImplementedError("Sequent hardware not implemented.")
+
+    def set_od(self, *args, **kwargs) -> None:
+        """Calls harware I/O or a simulated substitute depending
+        whether the class was instantiated with simulate = True.
+
```
Comment (tribeiro) [inline python/lsst/ts/vent/controller/controller.py:469]: Please, document input args.

File: python/lsst/ts/vent/controller/dispatcher.py [hunk 20]
```diff
@@ -27,7 +49,15 @@ def _cast_string_to_type(new_type: type, value: str):
     """
```
Comment (tribeiro) [inline python/lsst/ts/vent/controller/dispatcher.py:57]: Please document return value

**Review by tribeiro — APPROVED** [review]
I have a few additional comments I hope you will take into account but otherwise this looks good! thanks for the updates!

**Pull request merged**